### ISSUE TITLE
Djl Mxnet Engine train performance 6x+ slower than Python tensorflow 2.4.1

### ISSUE THREAD

**Issue by wangtieqiao** [deepjavalibrary/djl#932]
## Description
Not sure why Djl/Mxnet training is so much slower than Tensorflow, from what I read online, it should be other way around. I did a test with random data set, and result are:
Dataset:
- X (1_500_000, 500)
- y(1_500_000,1)
- batch size 1024x32
- Train on CPU for 10 epochs, with MLP hidden layers (1024, 128, 32)

Python tensorflow: elapsed **123** seconds 
while Djl/Mxnet elapsed **831** seconds.

## How to Reproduce?
# Python code:

      import os, sys
      os.environ["CUDA_VISIBLE_DEVICES"] = "-1"
      
      
      
      import numpy as np
      import tensorflow as tf
      import pandas as pd
      import datetime
      
      print("python version: ", sys.version)
      print("tensorflow version", tf.__version__)
      print("np version", np.__version__)
      
      nFeatures = 500
      nRecords = 1500000
      epoch = 10
      batchSize = 1024*32
      
      X = np.random.rand(nRecords, nFeatures)
      y = np.random.rand(nRecords, 1)
      print("X.shape: ", X.shape, "y.shape:", y.shape)
      
      model = tf.keras.Sequential([
          tf.keras.layers.Dense(1024, activation="relu"),
          tf.keras.layers.Dropout(rate=0.5),
          tf.keras.layers.Dense(128, activation="relu"),
          tf.keras.layers.Dropout(rate=0.5),
          tf.keras.layers.Dense(32, activation="relu"),
          tf.keras.layers.Dropout(rate=0.5),
          tf.keras.layers.Dense(1)
      ])
      optimizer = tf.keras.optimizers.Adam()
      loss = "mse"
      model.compile(optimizer, loss)
      
      ts = datetime.datetime.now()
      model.fit(X, y, batch_size=batchSize, epochs=epoch, shuffle=True)
      print("{} epochs elapsed {}".format(epoch, datetime.datetime.now() - ts))

# python package versions
      
       python version:  3.8.8 (default, Feb 24 2021, 21:46:12) 
        [GCC 7.3.0]
        tensorflow version 2.4.1
        numpy version 1.19.5

# Python output:

      2021-05-03 20:28:00.431163: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
      python version:  3.8.8 (default, Feb 24 2021, 21:46:12) 
      [GCC 7.3.0]
      tensorflow version 2.4.1
      np version 1.19.5
      X.shape:  (1500000, 500) y.shape: (1500000, 1)
      2021-05-03 20:28:09.850262: I tensorflow/compiler/jit/xla_cpu_device.cc:41] Not creating XLA devices, tf_xla_enable_xla_devices not set
      2021-05-03 20:28:09.851314: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcuda.so.1
      2021-05-03 20:28:09.875531: E tensorflow/stream_executor/cuda/cuda_driver.cc:328] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
      2021-05-03 20:28:09.875565: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:169] retrieving CUDA diagnostic information for host: tieqiao-ryzen9-3900x
      2021-05-03 20:28:09.875572: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:176] hostname: tieqiao-ryzen9-3900x
      2021-05-03 20:28:09.876001: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:200] libcuda reported version is: 460.73.1
      2021-05-03 20:28:09.876030: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:204] kernel reported version is: 460.73.1
      2021-05-03 20:28:09.876037: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:310] kernel version seems to match DSO: 460.73.1
      2021-05-03 20:28:09.876277: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
      To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
      2021-05-03 20:28:09.877289: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set
      2021-05-03 20:28:11.342499: W tensorflow/core/framework/cpu_allocator_impl.cc:80] Allocation of 3000000000 exceeds 10% of free system memory.
      2021-05-03 20:28:12.205332: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:116] None of the MLIR optimization passes are enabled (registered 2)
      2021-05-03 20:28:12.226011: I tensorflow/core/platform/profile_utils/cpu_utils.cc:112] CPU Frequency: 3792905000 Hz
      Epoch 1/10
      46/46 [==============================] - 13s 267ms/step - loss: 0.6052
      Epoch 2/10
      46/46 [==============================] - 12s 266ms/step - loss: 0.1410
      Epoch 3/10
      46/46 [==============================] - 12s 260ms/step - loss: 0.1189
      Epoch 4/10
      46/46 [==============================] - 12s 263ms/step - loss: 0.1108
      Epoch 5/10
      46/46 [==============================] - 12s 264ms/step - loss: 0.1056
      Epoch 6/10
      46/46 [==============================] - 12s 264ms/step - loss: 0.1009
      Epoch 7/10
      46/46 [==============================] - 12s 261ms/step - loss: 0.0971
      Epoch 8/10
      46/46 [==============================] - 12s 258ms/step - loss: 0.0942
      Epoch 9/10
      46/46 [==============================] - 12s 257ms/step - loss: 0.0921
      Epoch 10/10
      46/46 [==============================] - 12s 260ms/step - loss: 0.0909
      10 epochs elapsed 0:02:03.393369
      

# Kotlin/Djl/Mxnet source code
      import ai.djl.Device
      import ai.djl.Model
      import ai.djl.metric.Metrics
      import ai.djl.ndarray.NDList
      import ai.djl.ndarray.NDManager
      import ai.djl.ndarray.types.Shape
      import ai.djl.nn.Activation
      import ai.djl.nn.SequentialBlock
      import ai.djl.nn.core.Linear
      import ai.djl.nn.norm.Dropout
      import ai.djl.training.DefaultTrainingConfig
      import ai.djl.training.EasyTrain
      import ai.djl.training.dataset.Dataset
      import ai.djl.training.dataset.RandomAccessDataset
      import ai.djl.training.dataset.Record
      import ai.djl.training.listener.TrainingListener
      import ai.djl.training.loss.Loss
      import ai.djl.training.optimizer.Optimizer
      import ai.djl.training.tracker.Tracker
      import ai.djl.training.tracker.WarmUpTracker
      import ai.djl.util.Progress
      import kotlin.random.Random
      class FloatArrayListDataSet(builder: Builder) : RandomAccessDataset(builder) {
          val X: List<FloatArray> = builder.X!!
          val y: List<FloatArray>? = builder.y
          init {
              require(y == null || y.size == X.size) { "features does not have same number of rows as labels" }
          }
          class Builder() : BaseBuilder<Builder>() {
              var X: List<FloatArray>? = null
              var y: List<FloatArray>? = null
              var usage: Dataset.Usage = Dataset.Usage.TRAIN
              override fun self() = this
              fun build(): FloatArrayListDataSet {
                  require(X != null) { "features is not set" }
                  return FloatArrayListDataSet(this)
              }
              fun setX(X: List<FloatArray>): Builder {
                  this.X = X
                  return this
              }
              fun setY(y: List<FloatArray>): Builder {
                  this.y = y
                  return this
              }
          }
          override fun prepare(progress: Progress) {}
          override fun prepare() {}
          override fun get(manager: NDManager, index: Long): Record {
              val idx = index.toInt()
              val data = NDList(manager.create(X[idx]))
              val label = NDList()
              if (y != null)
                  label.add(manager.create(y[idx]))
              return Record(data, label)
          }
          override fun availableSize() = X.size.toLong()
          companion object {
              fun builder() = Builder()
          }
      }
      fun main() {
          val nFeatures = 500L
          val nRecords = 1_500_000
          val batchSize = 1024*32
          val X = (0 until nRecords).map {
              (0 until nFeatures).map { Random.nextFloat() }.toFloatArray()
          }
          val y = (0 until nRecords).map {
              (0 until 1).map { Random.nextFloat() }.toFloatArray()
          }
          println("X: ${X.size}x${X.first().size}, y: ${y.size}")
          val mlp = SequentialBlock()
              .add(Linear.builder().setUnits(1024).build())
              .add(Activation.reluBlock())
              .add(Dropout.builder().optRate(0.5f).build())
              .add(Linear.builder().setUnits(128).build())
              .add(Activation.reluBlock())
              .add(Dropout.builder().optRate(0.5f).build())
              .add(Linear.builder().setUnits(32).build())
              .add(Activation.reluBlock())
              .add(Dropout.builder().optRate(0.5f).build())
              .add(Linear.builder().setUnits(1).build())
          val optimizer = Optimizer.adam().build()
          val trainCfg = DefaultTrainingConfig(Loss.l2Loss())
              // try to set up Adam optomizer with clipGradidant 0.5f just a number came out of dark, try to see if cause any issue again
              .optOptimizer(optimizer)
              .addTrainingListeners(*TrainingListener.Defaults.logging())
              .optDevices(Device.getDevices(0))
          val ds = FloatArrayListDataSet.builder().setX(X).setY(y).setSampling(batchSize, true).build()
          val model = Model.newInstance("test").also { it.block = mlp }
          val trainer = model.newTrainer(trainCfg)
          val inputShape = Shape(1L, nFeatures)
          trainer!!.initialize(inputShape)
          val epoch = 10
          val ts = System.currentTimeMillis()
          EasyTrain.fit(trainer, epoch, ds, null)
          println("Traion ${epoch} epochs elapsed ${System.currentTimeMillis() - ts} milliseconds")
      }

# kotlin/djl/mxnet version 

    implementation("ai.djl:api:0.10.0")
    implementation("ai.djl.mxnet:mxnet-engine:0.10.0")
    implementation("ai.djl.mxnet:mxnet-model-zoo:0.10.0")
    //    implementation("ai.djl.mxnet:mxnet-native-auto:1.7.0-backport")
    implementation("ai.djl.mxnet:mxnet-native-auto:1.8.0")

# djl output

        Task :Djl_perfKt.main()
        X: 1500000x500, y: 1500000
        18:25:01,247 |-INFO in ch.qos.logback.classic.LoggerContext[default] - Could NOT find resource [logback-test.xml]
        18:25:01,248 |-INFO in ch.qos.logback.classic.LoggerContext[default] - Could NOT find resource [logback.groovy]
        18:25:01,248 |-INFO in ch.qos.logback.classic.LoggerContext[default] - Found resource [logback.xml] at [file:/home/tieqiao/git_repo/of/of_kt/build/resources/main/logback.xml]
        18:25:01,249 |-WARN in ch.qos.logback.classic.LoggerContext[default] - Resource [logback.xml] occurs multiple times on the classpath.
        18:25:01,249 |-WARN in ch.qos.logback.classic.LoggerContext[default] - Resource [logback.xml] occurs at [jar:file:/home/tieqiao/.gradle/caches/modules-2/files-2.1/com.tylerthrailkill.helpers/pretty-print/v2.0.8/462295aefa7d6aa5caea68e6560224806bc13e6a/pretty-print-v2.0.8.jar!/logback.xml]
        18:25:01,249 |-WARN in ch.qos.logback.classic.LoggerContext[default] - Resource [logback.xml] occurs at [file:/home/tieqiao/git_repo/of/of_kt/build/resources/main/logback.xml]
        18:25:01,313 |-INFO in ch.qos.logback.classic.joran.action.ConfigurationAction - debug attribute not set
        18:25:01,314 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - About to instantiate appender of type [ch.qos.logback.core.ConsoleAppender]
        18:25:01,318 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - Naming appender as [STDOUT]
        18:25:01,321 |-INFO in ch.qos.logback.core.joran.action.NestedComplexPropertyIA - Assuming default type [ch.qos.logback.classic.encoder.PatternLayoutEncoder] for [encoder] property
        18:25:01,334 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - About to instantiate appender of type [ch.qos.logback.core.helpers.NOPAppender]
        18:25:01,334 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - Naming appender as [NULL]
        18:25:01,334 |-INFO in ch.qos.logback.classic.joran.action.RootLoggerAction - Setting level of ROOT logger to INFO
        18:25:01,334 |-INFO in ch.qos.logback.core.joran.action.AppenderRefAction - Attaching appender named [STDOUT] to Logger[ROOT]
        18:25:01,335 |-INFO in ch.qos.logback.classic.joran.action.ConfigurationAction - End of configuration.
        18:25:01,335 |-INFO in ch.qos.logback.classic.joran.JoranConfigurator@780cb77 - Registering current configuration as safe fallback point
        18:25:02.904 [main] WARN  ai.djl.mxnet.jna.LibUtils - No matching cuda flavor for linux found: cu113mkl/sm_75.
        18:25:03.055 [main] INFO  a.d.t.l.LoggingTrainingListener - Training on: cpu().
        18:25:03.056 [main] INFO  a.d.t.l.LoggingTrainingListener - Load MXNet Engine Version 1.8.0 in 0.018 ms.
        Training:    100% |████████████████████████████████████████| 
        18:26:25.363 [main] INFO  a.d.t.l.LoggingTrainingListener - Epoch 1 finished.
        Training:    100% |████████████████████████████████████████| 
        18:27:48.993 [main] INFO  a.d.t.l.LoggingTrainingListener - Epoch 2 finished.
        Training:    100% |████████████████████████████████████████| 
        18:29:11.559 [main] INFO  a.d.t.l.LoggingTrainingListener - Epoch 3 finished.
        Training:    100% |████████████████████████████████████████| 
        18:30:35.951 [main] INFO  a.d.t.l.LoggingTrainingListener - Epoch 4 finished.
        Training:    100% |████████████████████████████████████████| 
        18:31:57.282 [main] INFO  a.d.t.l.LoggingTrainingListener - Epoch 5 finished.
        Training:    100% |████████████████████████████████████████| 
        18:33:20.540 [main] INFO  a.d.t.l.LoggingTrainingListener - Epoch 6 finished.
        Training:    100% |████████████████████████████████████████| 
        18:34:44.731 [main] INFO  a.d.t.l.LoggingTrainingListener - Epoch 7 finished.
        Training:    100% |████████████████████████████████████████| 
        18:36:08.984 [main] INFO  a.d.t.l.LoggingTrainingListener - Epoch 8 finished.
        Training:    100% |████████████████████████████████████████| 
        18:37:31.312 [main] INFO  a.d.t.l.LoggingTrainingListener - Epoch 9 finished.
        Training:    100% |████████████████████████████████████████| 
        18:38:54.734 [main] INFO  a.d.t.l.LoggingTrainingListener - Epoch 10 finished.
        Traion 10 epochs elapsed 831649 milliseconds

**Comment by wangtieqiao** [comment]
Looks might related to MxNet settings.

I saw tensorflow run with full cpu cores, while MxNet only use few of cores.
So believe might has nothing to do with the djl implementation.
